hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2008, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.operation;
+
+import com.hazelcast.config.MapConfig;
+import com.hazelcast.map.impl.MapDataSerializerHook;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.projection.Projection;
+import com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl;
+import com.hazelcast.spi.serialization.SerializationService;
+
+import java.io.IOException;
+
+import static com.hazelcast.config.InMemoryFormat.OBJECT;
+import static com.hazelcast.spi.ExecutionService.ASYNC_EXECUTOR;
+import static com.hazelcast.spi.properties.GroupProperty.GET_PROJECTION_OFFLOAD;
+import static com.hazelcast.spi.properties.GroupProperty.GET_PROJECTION_OFFLOAD_EXECUTOR_NAME;
+
+public final class GetWithProjectionOperation extends GetOperation {
+
+    private Projection projection;
+    private Data result;
+
+    public GetWithProjectionOperation() {
+    }
+
+    public GetWithProjectionOperation(String name, Data dataKey, Projection projection) {
+        super(name, dataKey);
+
+        this.dataKey = dataKey;
+        this.projection = projection;
+    }
+
+    @Override
+    public void run() {
+        Object value = recordStore.get(dataKey, false);
+
+        boolean offload = getNodeEngine().getProperties().getBoolean(GET_PROJECTION_OFFLOAD);
+        SerializationService ss = getNodeEngine().getSerializationService();
+        if (offload) {
+            value = isObjectFormat() ? ss.toObject(ss.toData(value)) : ss.toObject(value);
```

Comment:
I would move the toObject into the offloading. So you convert object -> data on the partition thread, but the offloading thread can take care of deserializing data to object. This removes more pressure for the partition thread and will allow for more parallel operations on the same partition.